### PR TITLE
Implement Keycloak authentication

### PR DESCRIPTION
**Pull request by lunasy26**

### CODE REVIEW

File: src/Presentation/Memo4me.UI/Program.cs [hunk 1]
```diff
@@ -1,35 +1,35 @@
 using Memo4me.UI.Components;
+using Memo4me.UI.Extensions;
 
-namespace Memo4me.UI;
-
-public class Program
+var builder = WebApplication.CreateBuilder(args); 
+{
+    var configuration = builder.Configuration;
```
Comment (Kharlap-Sergey) [inline src/Presentation/Memo4me.UI/Program.cs]: why can't we pass it directly?

File: src/Presentation/Memo4me.UI/Program.cs [hunk 2]
```diff
@@ -1,35 +1,35 @@
 using Memo4me.UI.Components;
+using Memo4me.UI.Extensions;
 
-namespace Memo4me.UI;
-
-public class Program
+var builder = WebApplication.CreateBuilder(args); 
+{
```
Comment (Kharlap-Sergey) [inline src/Presentation/Memo4me.UI/Program.cs]: Why do we need new code block here

File: src/Presentation/Memo4me.UI/Components/Pages/Logout.razor [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+<h3>Logout</h3>
+
+@code {
+    protected override async Task OnInitializedAsync()
+    {
+        var httpContext = HttpContextAccessor.HttpContext;
+        var idToken = await httpContext.GetTokenAsync("id_token");
+        await httpContext.SignOutAsync();
+        
+        var keycloakLogoutUrl = $"http://host.docker.internal:7270/realms/MemoRealm/protocol/openid-connect/logout" +
```
Comment (Kharlap-Sergey) [inline src/Presentation/Memo4me.UI/Components/Pages/Logout.razor]: hardcoded url

File: src/Presentation/Memo4me.UI/Components/Pages/Login.razor [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿@page "/login"
+@using Microsoft.AspNetCore.Authentication
+@rendermode InteractiveServer
+@inject NavigationManager UriHelper
+@inject IHttpContextAccessor HttpContextAccessor
+
+<h3>Login</h3>
+
+@code {
+    protected override async Task OnInitializedAsync()
```
Comment (Kharlap-Sergey) [inline src/Presentation/Memo4me.UI/Components/Pages/Login.razor:10]: Could you please explain to me, when and in which conditions this method is invoked

File: src/Presentation/Memo4me.UI/Extensions/AddAuthOpenIdExtension.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            })
+            .AddCookie()
+            .AddOpenIdConnect("oidc", options =>
+            {
+                options.Authority = configuration["Authentication:OpenIdConnect:Authority"];
+                options.ClientId = configuration["Authentication:OpenIdConnect:ClientId"];
+                options.ClientSecret = configuration["Authentication:OpenIdConnect:ClientSecret"];
+                options.ResponseType = configuration["Authentication:OpenIdConnect:ResponseType"];
+                options.SaveTokens = true;
+                options.RequireHttpsMetadata = false;
```
Comment (Kharlap-Sergey) [inline src/Presentation/Memo4me.UI/Extensions/AddAuthOpenIdExtension.cs]: should be configurable, and set to true by default, as it critical for production

File: src/Presentation/Memo4me.UI/appsettings.json
Comment (Kharlap-Sergey) [inline src/Presentation/Memo4me.UI/appsettings.json:1]: General recommendation
appsettings.json should contain schema with values required for production, if you need something for development use appsettings.Dev...json 
Secret information shouldn't be stored here at all
When run in VS there is UserSecrets configuration provider, but when run through docker you don't have such an option so it better to provider them through env variables

File: src/Presentation/Memo4me.UI/Program.cs (hunk 2 above)
Comment (lunasy26) [reply src/Presentation/Memo4me.UI/Program.cs]: for visual convenience
If this is wrong, I'll fix it

File: src/Presentation/Memo4me.UI/Components/Pages/Login.razor (hunk 4 above)
Comment (lunasy26) [reply src/Presentation/Memo4me.UI/Components/Pages/Login.razor:10]: OnInitializedAsync() is a lifecycle method of the blazor component that is called once when the component is initialized and after it has been added to rendering
In my case i use this method to initialize the authentication process, sending the user to the login page

File: src/Presentation/Memo4me.UI/Extensions/AddAuthOpenIdExtension.cs (hunk 5 above)
Comment (lunasy26) [reply src/Presentation/Memo4me.UI/Extensions/AddAuthOpenIdExtension.cs]: done

File: src/Presentation/Memo4me.UI/Components/Pages/Logout.razor (hunk 3 above)
Comment (lunasy26) [reply src/Presentation/Memo4me.UI/Components/Pages/Logout.razor]: fixed

File: src/Presentation/Memo4me.UI/Program.cs (hunk 2 above)
Comment (Kharlap-Sergey) [reply src/Presentation/Memo4me.UI/Program.cs]: It is ok, I just never seen this so it baffles the eye

**Review by Kharlap-Sergey — APPROVED** [review]

**Pull request merged**